### ISSUE TITLE
[Feature] Get the real logprobs to analyze decoding

### ISSUE THREAD

**Issue by Snowdar** [sgl-project/sglang#2178]
### Checklist

- [X] 1. If the issue you raised is not a feature but a question, please raise a discussion at https://github.com/sgl-project/sglang/discussions/new/choose Otherwise, it will be closed.
- [X] 2. Please use English, otherwise it will be closed.

### Motivation

Can logprobs return probabilities after sampling processing? Currently, the logprobs seem to be the probability distribution before processing such as temperature and top-p truncation.

**Comment by yhyang201** [comment]
I attempted to implement this feature, but I encountered an issue:

When the sampling backend is set to `flashinfer`, the function `min_p_sampling_from_probs` within `flashinfer` is called for sampling. However, `sglang` is unable to access the `probs` values processed by the `min_p` operation (i.e., the `min_p`-processed `probs`). 

To obtain the processed probs values, it will be necessary to compute the `min_p`-processed `probs` values separately when using `flashinfer` as the sampling backend. This may be a bit challenging, as the implementation of `flashinfer.min_p_sampling_from_probs` appears to be non-trivial:  
[MinPSamplingFromProbKernel](https://github.com/flashinfer-ai/flashinfer/blob/5fe9f7d1d1ab8aa13cb6073a6447e383ad52b484/include/flashinfer/sampling.cuh#L751)].

**Comment by Snowdar** [comment]
> I attempted to implement this feature, but I encountered an issue:
> 
> When the sampling backend is set to `flashinfer`, the function `min_p_sampling_from_probs` within `flashinfer` is called for sampling. However, `sglang` is unable to access the `probs` values processed by the `min_p` operation (i.e., the `min_p`-processed `probs`).
> 
> To obtain the processed probs values, it will be necessary to compute the `min_p`-processed `probs` values separately when using `flashinfer` as the sampling backend. This may be a bit challenging, as the implementation of `flashinfer.min_p_sampling_from_probs` appears to be non-trivial: [MinPSamplingFromProbKernel](https://github.com/flashinfer-ai/flashinfer/blob/5fe9f7d1d1ab8aa13cb6073a6447e383ad52b484/include/flashinfer/sampling.cuh#L751)].

Yes, there are many dependent functions that need to be modified here. If it is not the code of sglang itself, it will be a bit troublesome. Another way is to add a parameter to control whether to obtain the probability distribution before or after sampling, and then assume that the probability distribution after obtaining is generally used for analysis. Without considering response speed, we can use the implementation of torch to run it again, but we also need to align various implementations.

**Comment by yhyang201** [comment]
@Snowdar 
Hi, I have implemented this feature in my fork, which you can find at [https://github.com/Mayflyyh/sglang/tree/feature/2178-logprobs-post-sampling. However, I haven't had the chance to add unit tests or update the documentation yet, so I'm unable to submit a PR at the moment. But perhaps you could take a look in the meantime?
example: 
```
response = requests.post(
    "http://localhost:30000/generate",
    json={
        "text": "The capital of France is",
        "sampling_params": {
            "temperature": 0,
            # "max_new_tokens": 32,
        },
        "return_logprob": True,
        "return_sampled_prob": True
    },
)
```

**Comment by Snowdar** [comment]
> @Snowdar Hi, I have implemented this feature in my fork, which you can find at [https://github.com/Mayflyyh/sglang/tree/feature/2178-logprobs-post-sampling. However, I haven't had the chance to add unit tests or update the documentation yet, so I'm unable to submit a PR at the moment. But perhaps you could take a look in the meantime? example:
> 
> ```
> response = requests.post(
>     "http://localhost:30000/generate",
>     json={
>         "text": "The capital of France is",
>         "sampling_params": {
>             "temperature": 0,
>             # "max_new_tokens": 32,
>         },
>         "return_logprob": True,
>         "return_sampled_prob": True
>     },
> )
> ```

Good jobs! Thanks for contribute it.

By the way, to avoid change too many files and functions so that there is no hidden bugs and compatibility with official codes (expecially when the output of a func is changed), I have implemented it carefully by this way:
```
        # sglang/python/sglang/srt/model_executor/model_runner.py: sample function
        # Sample the next tokens.
        next_token_ids, probs = self.sampler(logits, sampling_info)
        if forward_batch.return_logprob and probs is not None:
            # get real logprobs after temperature scaling
            logprobs = probs.log().clamp(-1e10)
            logits_metadata = LogitsMetadata.from_forward_batch(forward_batch)
            input_top_logprobs, output_top_logprobs = self.model.logits_processor.get_top_logprobs(logprobs, logits_metadata)
            logits_output.next_token_logprobs = logprobs
            logits_output.input_top_logprobs = input_top_logprobs
            logits_output.output_top_logprobs = output_top_logprobs
```
The probs will be re-computed after sampling, but only the probs with temperature applied could be obtained (No top-p, top-k etc.)

**Comment by merrymercy** [comment]
Fixed by https://github.com/sgl-project/sglang/pull/2664

**Comment by rahul-sarvam** [comment]
@merrymercy thanks for this! I have a couple of questions:

1. how can I get the top-k logprobs for each generation step, instead of just getting the top-most/sampled token logprob?
2. is there a way to get the raw logits instead of the logprobs?

**Comment by winglian** [comment]
I'm also interested in getting the raw-logits/top-k.
